// @metaplex-foundation/amman-client/dist/utils/log.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.scopedLog = exports.logInfo = exports.logError = exports.logTrace = exports.logDebug = exports.logInfoDebug = exports.logErrorDebug = void 0;
const debug_1 = __importDefault(require("debug"));
exports.logErrorDebug = (0, debug_1.default)('amman:error');
exports.logInfoDebug = (0, debug_1.default)('amman:info');
exports.logDebug = (0, debug_1.default)('amman:debug');
exports.logTrace = (0, debug_1.default)('amman:trace');
exports.logError = exports.logErrorDebug.enabled
    ? exports.logErrorDebug
    : console.error.bind(console);
exports.logInfo = exports.logInfoDebug.enabled
    ? exports.logInfoDebug
    : console.log.bind(console);
function scopedLog(scope) {
    const logError = (0, debug_1.default)(`amman:${scope}:error`);
    const logInfo = (0, debug_1.default)(`amman:${scope}:info`);
    return {
        logError: logError.enabled ? logError : console.error.bind(console),
        logInfo: logInfo.enabled ? logInfo : console.log.bind(console),
        logDebug: (0, debug_1.default)(`amman:${scope}:debug`),
        logTrace: (0, debug_1.default)(`amman:${scope}:trace`),
    };
}
exports.scopedLog = scopedLog;
