// @metaplex-foundation/amman/dist/utils/config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fullAccountsDir = exports.completeConfig = exports.DEFAULT_START_CONFIG = exports.DEFAULT_STREAM_TRANSACTION_LOGS = void 0;
const path_1 = __importDefault(require("path"));
const assets_1 = require("../assets");
const types_1 = require("../relay/types");
const types_2 = require("../storage/types");
const validator_1 = require("../validator");
exports.DEFAULT_STREAM_TRANSACTION_LOGS = process.env.CI == null;
exports.DEFAULT_START_CONFIG = {
    validator: validator_1.DEFAULT_VALIDATOR_CONFIG,
    relay: types_1.DEFAULT_RELAY_CONFIG,
    snapshot: assets_1.DEFAULT_SNAPSHOT_CONFIG,
    storage: types_2.DEFAULT_STORAGE_CONFIG,
    streamTransactionLogs: exports.DEFAULT_STREAM_TRANSACTION_LOGS,
    assetsFolder: assets_1.DEFAULT_ASSETS_FOLDER,
};
/**
 * Need a hand rolled version of this since tools like deep-extend don't work
 * with maps and classes and thus break things like account renderers
 *
 * @private
 */
function completeConfig(config = {}) {
    const relay = { ...types_1.DEFAULT_RELAY_CONFIG, ...config.relay };
    const validator = {
        ...validator_1.DEFAULT_VALIDATOR_CONFIG,
        ...config.validator,
    };
    const snapshot = { ...assets_1.DEFAULT_SNAPSHOT_CONFIG, ...config.snapshot };
    const storage = { ...types_2.DEFAULT_STORAGE_CONFIG, ...config.storage };
    snapshot.snapshotFolder = path_1.default.resolve(process.cwd(), snapshot.snapshotFolder);
    const { streamTransactionLogs = exports.DEFAULT_STREAM_TRANSACTION_LOGS, assetsFolder = assets_1.DEFAULT_ASSETS_FOLDER, } = config;
    return {
        validator,
        relay,
        storage,
        snapshot,
        streamTransactionLogs,
        assetsFolder,
    };
}
exports.completeConfig = completeConfig;
function fullAccountsDir(assetsFolder) {
    const assetsDir = path_1.default.resolve(process.cwd(), assetsFolder !== null && assetsFolder !== void 0 ? assetsFolder : assets_1.DEFAULT_ASSETS_FOLDER);
    return path_1.default.join(assetsDir, assets_1.ACCOUNTS_FOLDER);
}
exports.fullAccountsDir = fullAccountsDir;
