// @metaplex-foundation/amman-client/dist/transactions/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfirmedTransactionDetails = void 0;
const asserts_1 = require("../asserts/asserts");
/**
 * Result returned by {@link TransactionHandler#sendAndConfirmTransaction}.
 *
 * @property txSignature {@link TransactionSignature} string of sent transaction
 * @property txRpcResponse initial response of sent transaction
 * @property txConfirmed the result of confirming the transaction
 * @property txSummary a summary of the confirmed transaction
 *
 * @category transactions
 */
class ConfirmedTransactionDetails {
    constructor(args) {
        this.txSignature = args.txSignature;
        this.txRpcResponse = args.txRpcResponse;
        this.txConfirmed = args.txConfirmed;
        this.txSummary = args.txSummary;
    }
    /**
     * Call this if to assert that the log messages match a given set of regular expressions.
     *
     * @param msgRxs it is verified that the logs match all these {@link RegExp}es
     * @param opts options to customize the assertion diagnostics
     */
    async assertLogs(t, msgRxs, opts) {
        (0, asserts_1.assertContainMessages)(t, this.txSummary.logMessages, msgRxs, opts, 'log messages');
    }
}
exports.ConfirmedTransactionDetails = ConfirmedTransactionDetails;
