// @metaplex-foundation/amman-client/dist/relay/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KILL_AMMAN_EXIT_CODE = exports.AmmanRequests = exports.MSG_RESPOND_KILL_AMMAN = exports.MSG_REQUEST_KILL_AMMAN = exports.MSG_RESPOND_VALIDATOR_PID = exports.MSG_REQUEST_VALIDATOR_PID = exports.MSG_RESPOND_AMMAN_VERSION = exports.MSG_REQUEST_AMMAN_VERSION = exports.MSG_RESPOND_RESTART_VALIDATOR = exports.MSG_REQUEST_RESTART_VALIDATOR = exports.MSG_RESPOND_LOAD_SNAPSHOT = exports.MSG_REQUEST_LOAD_SNAPSHOT = exports.MSG_RESPOND_SET_ACCOUNT = exports.MSG_REQUEST_SET_ACCOUNT = exports.MSG_RESPOND_LOAD_KEYPAIR = exports.MSG_REQUEST_LOAD_KEYPAIR = exports.MSG_RESPOND_STORE_KEYPAIR = exports.MSG_REQUEST_STORE_KEYPAIR = exports.MSG_RESPOND_ACCOUNT_SAVE = exports.MSG_REQUEST_ACCOUNT_SAVE = exports.MSG_RESPOND_SNAPSHOT_SAVE = exports.MSG_REQUEST_SNAPSHOT_SAVE = exports.MSG_RESPOND_ACCOUNT_STATES = exports.MSG_REQUEST_ACCOUNT_STATES = exports.MSG_UPDATE_ACCOUNT_STATES = exports.MSG_GET_KNOWN_ADDRESS_LABELS = exports.MSG_CLEAR_TRANSACTIONS = exports.MSG_CLEAR_ADDRESS_LABELS = exports.ACK_UPDATE_ADDRESS_LABELS = exports.MSG_UPDATE_ADDRESS_LABELS = exports.AMMAN_RELAY_URI = exports.AMMAN_RELAY_PORT = void 0;
/** @private */
exports.AMMAN_RELAY_PORT = 50474;
/** @private */
exports.AMMAN_RELAY_URI = `http://localhost:${exports.AMMAN_RELAY_PORT}`;
/** @private */
exports.MSG_UPDATE_ADDRESS_LABELS = 'update:address-labels';
/** @private */
exports.ACK_UPDATE_ADDRESS_LABELS = 'ack:update:address-labels';
/** @private */
exports.MSG_CLEAR_ADDRESS_LABELS = 'clear:address-labels';
/** @private */
exports.MSG_CLEAR_TRANSACTIONS = 'clear:transactions';
/** @private */
exports.MSG_GET_KNOWN_ADDRESS_LABELS = 'get:known-address-labels';
/** @private */
exports.MSG_UPDATE_ACCOUNT_STATES = 'update:account-states';
/** @private */
exports.MSG_REQUEST_ACCOUNT_STATES = 'request:account-states';
/** @private */
exports.MSG_RESPOND_ACCOUNT_STATES = 'respond:account-states';
/** @private */
exports.MSG_REQUEST_SNAPSHOT_SAVE = 'request:snapshot-save';
/** @private */
exports.MSG_RESPOND_SNAPSHOT_SAVE = 'respond:snapshot-save';
/** @private */
exports.MSG_REQUEST_ACCOUNT_SAVE = 'request:account-save';
/** @private */
exports.MSG_RESPOND_ACCOUNT_SAVE = 'respond:account-save';
/** @private */
exports.MSG_REQUEST_STORE_KEYPAIR = 'request:store-keypair';
/** @private */
exports.MSG_RESPOND_STORE_KEYPAIR = 'respond:store-keypair';
/** @private */
exports.MSG_REQUEST_LOAD_KEYPAIR = 'request:load-keypair';
/** @private */
exports.MSG_RESPOND_LOAD_KEYPAIR = 'respond:load-keypair';
/** @private */
exports.MSG_REQUEST_SET_ACCOUNT = 'request:set-account';
/** @private */
exports.MSG_RESPOND_SET_ACCOUNT = 'respond:set-account';
/** @private */
exports.MSG_REQUEST_LOAD_SNAPSHOT = 'request:load-snapshot';
/** @private */
exports.MSG_RESPOND_LOAD_SNAPSHOT = 'respond:load-snapshot';
/** @private */
exports.MSG_REQUEST_RESTART_VALIDATOR = 'request:restart-validator';
/** @private */
exports.MSG_RESPOND_RESTART_VALIDATOR = 'respond:restart-validator';
/** @private */
exports.MSG_REQUEST_AMMAN_VERSION = 'request:relay-version';
/** @private */
exports.MSG_RESPOND_AMMAN_VERSION = 'respond:relay-version';
/** @private */
exports.MSG_REQUEST_VALIDATOR_PID = 'request:validator-pid';
/** @private */
exports.MSG_RESPOND_VALIDATOR_PID = 'respond:validator-pid';
/** @private */
exports.MSG_REQUEST_KILL_AMMAN = 'request:kill-amman';
/** @private */
exports.MSG_RESPOND_KILL_AMMAN = 'respond:kill-amman';
exports.AmmanRequests = [
    exports.MSG_UPDATE_ADDRESS_LABELS,
    exports.MSG_GET_KNOWN_ADDRESS_LABELS,
    exports.MSG_REQUEST_ACCOUNT_STATES,
    exports.MSG_REQUEST_SNAPSHOT_SAVE,
    exports.MSG_REQUEST_ACCOUNT_SAVE,
    exports.MSG_REQUEST_STORE_KEYPAIR,
    exports.MSG_REQUEST_LOAD_KEYPAIR,
    exports.MSG_REQUEST_SET_ACCOUNT,
    exports.MSG_REQUEST_LOAD_SNAPSHOT,
    exports.MSG_REQUEST_RESTART_VALIDATOR,
    exports.MSG_REQUEST_AMMAN_VERSION,
    exports.MSG_REQUEST_VALIDATOR_PID,
    exports.MSG_REQUEST_KILL_AMMAN,
];
exports.KILL_AMMAN_EXIT_CODE = 0xbad;
