// @metaplex-foundation/amman-client/dist/storage/sdk-types.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toBasisPoints = exports.sol = exports.lamports = exports.amount = exports.SOL = void 0;
const web3_js_1 = require("@solana/web3.js");
const bn_js_1 = __importDefault(require("bn.js"));
exports.SOL = {
    symbol: 'SOL',
    decimals: 9,
};
// -----------------
// Amount Helpers
// -----------------
const amount = (basisPoints, currency) => {
    return {
        basisPoints: (0, exports.toBasisPoints)(basisPoints),
        currency,
    };
};
exports.amount = amount;
const lamports = (lamports) => {
    return (0, exports.amount)(lamports, exports.SOL);
};
exports.lamports = lamports;
const sol = (sol) => {
    return (0, exports.lamports)(sol * web3_js_1.LAMPORTS_PER_SOL);
};
exports.sol = sol;
const toBasisPoints = (value) => {
    return new bn_js_1.default(value);
};
exports.toBasisPoints = toBasisPoints;
